### PR TITLE
Fix/lp 1927721

### PR DESCRIPTION
**Pull request by ycliuhw**
We should never expect to upgrade resources when we refresh an application to use a local charm and those resources were not provided via --resource;
So this change allows us to just provide the resources we want to override but no need to provided all the resources via --resources when we run juju refresh;

## Checklist

 - [ ] ~Requires a [pylibjuju](https://github.com/juju/python-libjuju) change~
 - [ ] ~Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR~
 - [ ] ~Added or updated [doc.go](https://discourse.jujucharms.com/t/readme-in-packages/451) related to packages changed~
 - [x] Comments answer the question of why design decisions were made

## QA steps

```console
$ juju add-model t1 -c k1 && juju deploy ch:kelvin-snappass-test -m k1:t1

$ juju list-resources kelvin-snappass-test
Resource        Supplied by  Revision
redis-image     charmstore   1
snappass-image  charmstore   1
test-file       charmstore   3

$ juju refresh kelvin-snappass-test --path ./kelvin-snappass-test.charm

$ juju list-resources kelvin-snappass-test
Resource        Supplied by  Revision
redis-image     charmstore   1
snappass-image  charmstore   1
test-file       charmstore   3

$ juju refresh kelvin-snappass-test --path ./kelvin-snappass-test.charm --resource redis-image=redis

$ juju list-resources kelvin-snappass-test
Resource        Supplied by  Revision
redis-image     admin        2021-19-05T22:44  << upgraded
snappass-image  charmstore   1
test-file       charmstore   3

# see if there are any errors from the model log or juju status
$ juju debug-log -m k1:t1 --color --replay

$ juju status --relations --color --storage -m k1:t1

$ yml2json ./metadata.yaml | jq .resources
{
  "snappass-image": {
    "type": "oci-image",
    "description": "Docker image for SnapPass"
  },
  "a-new-resource": {
    "type": "oci-image",
    "description": "test resource"
  },
  "redis-image": {
    "type": "oci-image",
    "description": "Docker image for Redis"
  },
  "test-file": {
    "type": "file",
    "description": "test resource",
    "filename": "test.txt"
  }
}

$ juju refresh kelvin-snappass-test --path ./kelvin-snappass-test.charm
Added  charm "kelvin-snappass-test", revision 0, to the model
ERROR new resource "a-new-resource" was missing, please provide it via --resource

$ juju refresh kelvin-snappass-test --path ./kelvin-snappass-test.charm --resource a-new-resource=mysql
Added  charm "kelvin-snappass-test", revision 6, to the model

$ juju list-resources kelvin-snappass-test
Resource        Supplied by  Revision
a-new-resource  admin        2021-19-05T22:45
redis-image     admin        2021-19-05T22:44
snappass-image  charmstore   1
test-file       charmstore   3

```

## Documentation changes

No

## Bug reference

https://bugs.launchpad.net/juju/+bug/1928043

### CODE REVIEW

**Comment by ycliuhw** [comment]
[1928043 Resource error not exposed for OCI images during refresh charm](https://trello.com/c/H0zG5JVR/1591-resource-error-not-exposed-for-oci-images-during-refresh-charm)

File: cmd/juju/application/deployer/bundlehandler.go [hunk 1]
```diff
@@ -1178,7 +1178,7 @@ func (h *bundleHandler) upgradeCharm(change *bundlechanges.UpgradeCharmChange) e
 	if err != nil {
 		return errors.Trace(err)
 	}
-	filtered, err := utils.GetUpgradeResources(resourceLister, p.Application, resources, meta)
+	filtered, err := utils.GetUpgradeResources(chID.URL, resourceLister, p.Application, resources, meta)
```
Comment (wallyworld) [inline cmd/juju/application/deployer/bundlehandler.go:1181]: maybe just use `curl` as that what `chID.URL` was set to above

File: cmd/juju/application/utils/utils.go [hunk 2]
```diff
@@ -98,40 +99,55 @@ func getResources(
...
+		logger.Tracef("%q provided to upgrade existing resource", res.Name)
+		return true, nil
 	}
+
 	cur, ok := current[res.Name]
 	if !ok {
-		// If there's no information on the server, there should be.
-		return true
+		// If there's no information on the server, there might be a new resource added to the charm.
+		if newCharmURL.Schema == "local" {
```
Comment (wallyworld) [inline cmd/juju/application/utils/utils.go:136]: we have a `OriginLocal` const

**Review by wallyworld — APPROVED** [review]
Great, nice to see this fixed

**Comment by ycliuhw** [comment]
$$merge$$

**Pull request merged**